### PR TITLE
Service layer for tenant wise notification sender configuration

### PR DESCRIPTION
**Pull request by senthalan**
### Proposed changes in this pull request

Service layer for tenant-wise notification sender configuration. 

Move the core business logic from https://github.com/wso2/identity-api-server/blob/v1.1.15/components/org.wso2.carbon.identity.api.server.notification.sender/org.wso2.carbon.identity.api.server.notification.sender.v1/src/main/java/org/wso2/carbon/identity/api/server/notification/sender/v1/core/NotificationSenderManagementService.java. (Will be remove with https://github.com/wso2/identity-api-server/pull/368 )

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+
+            reDeployEventPublisherConfiguration(smsSenderResource);
+        } catch (ConfigurationManagementException e) {
+            throw handleConfigurationMgtException(e, ERROR_CODE_ERROR_UPDATING_NOTIFICATION_SENDER,
+                    smsSender.getName());
+        }
+        return buildSmsSenderFromResource(smsSenderResource);
+    }
+
+    private Resource getPublisherResource(String resourceName) throws NotificationSenderManagementException {
```
Comment (chamathns) [inline components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementServiceImpl.java:330]: shall we use Optional<Resource> here?

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+                NotificationSenderManagementConstants.PUBLISHER_RESOURCE_TYPE, senderName,
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId());
+
+        int numberOfRetries = 0;
+        if (log.isDebugEnabled()) {
+            log.debug("Sending cluster invalidation message to other cluster nodes for event publisher delete for "
+                    + senderName);
+        }
+
+        while (numberOfRetries < 60) {
```
Comment (chamathns) [inline components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementServiceImpl.java:375]: Shall we use a constant for this value?

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/utils/NotificationSenderUtils.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+            attribute.setValue(property.getKey());
+            adapterProperty.setAttributeNode(attribute);
+            adapterProperty.appendChild(document.createTextNode(property.getValue()));
+            to.appendChild(adapterProperty);
+        }
+    }
+
+    private static void addToElementToSMSEventPublisher(SMSSenderDTO smsSender, Map<String, String> properties,
+                                                        Document document,
+                                                        Element root) {
```
Comment (chamathns) [inline components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/utils/NotificationSenderUtils.java:308]: Fix formatting

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementConstants.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        ERROR_CODE_CONFLICT_PUBLISHER("60002", "Notification sender already exists.",
+                "There exists a notification sender: %s in the tenant."),
+        ERROR_CODE_SMS_PROVIDER_REQUIRED("60003", "Required attribute is missing",
+                "SMS provider is not defined for notification sender."),
+        ERROR_CODE_SMS_PAYLOAD_NOT_FOUND("60004", "SMS send API payload is not defined",
+                "SMS send API payload for provider: %s, is not defined in file or POST body properties."),
+        ERROR_CODE_SMS_PROVIDER_URL_REQUIRED("60005", "Required attribute is missing",
+                "SMS provider url is not defined for notification sender."),
+        ERROR_CODE_PUBLISHER_NOT_EXISTS("60006", "No notification sender found.",
+                "No notification sender found : %s"),
```
Comment (IsuraD) [inline components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementConstants.java:130]: All these are input validations.. why are these not consistent?

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementService.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * Update email sender details.
+     *
+     * @param emailSender Email sender's updated configurations.
+     * @return Updated email sender.
+     * @throws NotificationSenderManagementException    Notification sender management exception.
+     */
+    EmailSenderDTO updateEmailSender(EmailSenderDTO emailSender) throws NotificationSenderManagementException;
+
+    /**
+     * Update sms sender details by name.
```
Comment (IsuraD) [inline components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementService.java:102]: Update sms sender details

File: components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementConstants.java (hunk 4 above)
Comment (senthalan) [reply components/notification-sender-config/org.wso2.carbon.identity.notification.sender.tenant.config/src/main/java/org/wso2/carbon/identity/notification/sender/tenant/config/NotificationSenderManagementConstants.java:130]: These error codes are already defined in  https://github.com/wso2/identity-api-server/pull/368/files#diff-8ef58086d9a4911e6c1e10b51a8b6c80892035b11d7bb3d2859bb64784078789L114-L124.

**Pull request merged**